pkg/providers/tf/job_runner.go

```diff
@@ -0,0 +1,233 @@
+	go func() {
+		err := workspace.Apply()
+		runner.operationFinished(err, workspace, deployment)
+	}()
+
+	return nil
+}
+
+// Destroy runs `terraform destroy` on the given workspace in the background.
+// The status of the job can be found by polling the Status function.
+func (runner *TfJobRunner) Destroy(ctx context.Context, id string) error {
+	deployment, err := db_service.GetTerraformDeploymentById(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	workspace, err := runner.hydrateWorkspace(ctx, deployment)
+	if err != nil {
+		return err
+	}
+
+	if err := runner.markJobStarted(ctx, deployment, models.DeprovisionOperationType); err != nil {
+		return err
+	}
+
+	go func() {
+		err := workspace.Destroy()
+		runner.operationFinished(err, workspace, deployment)
+	}()
+
+	return nil
+}
+
+// operationFinished closes out the state of the background job so clients that
+// are polling can get the results.
+func (runner *TfJobRunner) operationFinished(err error, workspace *wrapper.TerraformWorkspace, deployment *models.TerraformDeployment) error {
+	if err == nil {
+		deployment.LastOperationState = Succeeded
+		deployment.LastOperationMessage = ""
+	} else {
+		deployment.LastOperationState = Failed
+		deployment.LastOperationMessage = err.Error()
+	}
+
+	workspaceString, err := workspace.Serialize()
+	if err != nil {
+		deployment.LastOperationState = Failed
+		deployment.LastOperationMessage = fmt.Sprintf("couldn't serialize workspace, contact your operator for cleanup: %s", err.Error())
+	}
+
+	deployment.Workspace = workspaceString
+
+	return db_service.SaveTerraformDeployment(context.Background(), deployment)
+}
+
+// Status gets the status of the most recent job on the workspace.
+// If isDone is true, then the status of the operation will not change again.
+// if isDone is false, then the operation is ongoing.
+func (runner *TfJobRunner) Status(ctx context.Context, id string) (isDone bool, err error) {
+	deployment, err := db_service.GetTerraformDeploymentById(ctx, id)
+	if err != nil {
+		return true, err
+	}
+
+	switch deployment.LastOperationState {
+	case Succeeded:
+		return true, nil
+	case Failed:
+		return true, errors.New(deployment.LastOperationMessage)
+	default:
+		return false, nil
+	}
+}
+
+// Outputs gets the output variables for the given module instance in the workspace.
+func (runner *TfJobRunner) Outputs(ctx context.Context, id, instanceName string) (map[string]interface{}, error) {
+	deployment, err := db_service.GetTerraformDeploymentById(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	ws, err := wrapper.DeserializeWorkspace(deployment.Workspace)
+	if err != nil {
+		return nil, err
+	}
+
+	return ws.Outputs(instanceName)
+}
+
+// Wait waits for an operation to complete, polling its status once per second.
+func (runner *TfJobRunner) Wait(ctx context.Context, id string) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-time.After(1 * time.Second):
+			isDone, err := runner.Status(ctx, id)
```

Comment:
Will the runner time out on its own, or do we need to think about implementing that here to prevent infinite waiting?